mesa_geo/geospace.py

```diff
@@ -78,45 +102,145 @@ def add_agents(self, agents, auto_convert_crs=False):
-                    if auto_convert_crs:
-                        agent.to_crs(self.crs)
-                    else:
-                        raise ValueError(
-                            f"Inconsistent crs: {agent.__class__.__name__} is of crs {agent.crs.to_string()}, "
-                            f"different from the crs of {self.__class__.__name__} - {self.crs.to_string()}. "
-                            "Please check your crs settings, or set `auto_convert_crs` to `True` to allow "
-                            "automatic crs conversion of GeoAgents to GeoSpace."
-                        )
-            self._recreate_rtree(agents)
+                self._check_agent(agent, auto_convert_crs)
+        self._agent_layer.add_agents(agents)
+
+    def _recreate_rtree(self, new_agents=None):
+        """Create a new rtree index from agents geometries."""
+        self._agent_layer._recreate_rtree(new_agents)
+
+    def remove_agent(self, agent):
+        """Remove an agent from the GeoSpace."""
+        self._agent_layer.remove_agent(agent)
+
+    def get_relation(self, agent, relation):
+        """Return a list of related agents.
+
+        Args:
+            agent: the agent for which to compute the relation
+            relation: must be one of 'intersects', 'within', 'contains',
+                'touches'
+            other_agents: A list of agents to compare against.
+                Omit to compare against all other agents of the GeoSpace
+        """
+        yield from self._agent_layer.get_relation(agent, relation)
+
+    def get_intersecting_agents(self, agent, other_agents=None):
+        return self._agent_layer.get_intersecting_agents(agent, other_agents)
+
+    def get_neighbors_within_distance(
+        self, agent, distance, center=False, relation="intersects"
+    ):
+        """Return a list of agents within `distance` of `agent`.
+
+        Distance is measured as a buffer around the agent's geometry,
+        set center=True to calculate distance from center.
+        """
+        yield from self._agent_layer.get_neighbors_within_distance(
+            agent, distance, center, relation
+        )
+
+    def agents_at(self, pos):
+        """Return a list of agents at given pos."""
+        return self._agent_layer.agents_at(pos)
+
+    def distance(self, agent_a, agent_b):
+        """Return distance of two agents."""
+        return self._agent_layer.distance(agent_a, agent_b)
+
+    def get_neighbors(self, agent):
+        """Get (touching) neighbors of an agent."""
+        return self._agent_layer.get_neighbors(agent)
+
+
+class AgentLayer:
+    """Layer that contains the GeoAgents. Mainly for internal usage within `GeoSpace`.
+
+    Properties:
+        idx: R-tree index for fast spatial queries
+        bounds: Bounds of the layer in [min_x, max_x, min_y, max_y] format
+        agents: List of all agents in the layer
+
+    Methods:
+        add_agents: add a list or a single GeoAgent.
```

Comment:
I think there should also be a documentation that tells people that `add_agents` doesn't do the `_check_agent` stuff. Not sure of the best place to put it. Maybe in the docstring of `add_agents` itself.